Initial commit of PrusaColorChange plugin

dmwatson opened the pull request:
This plugin inserts M600 filament change G-code into the requested layer or layers of a print. Optionally, the filament ejection amount can be changed also to fit different types of extruders.

dmwatson committed:
Initial commit of PrusaColorChange plugin

nallath commented:
For our reference; CURA-4326

Ghostkeeper commented:
I hope you don't mind if we clean up the style of the code a bit? The code itself is very simple but there's all sorts of draft code in between and the code style is inconsistent.

Ghostkeeper commented on scripts/PrusaColorChange.py:
```diff
@@ -0,0 +1,95 @@
+                    "description": "Later filament retraction distance for removal",
+                    "unit": "mm",
+                    "type": "float",
+                    "default_value": 3000.0
```
Are you sure that a retraction distance of 3000mm is correct? That's 3 metres!

dmwatson replied:
That's a bit off. I think I meant 30.0 as the default for extra.

Ghostkeeper commented on scripts/PrusaColorChange.py:
```diff
@@ -0,0 +1,95 @@
+                layer_num = int( layer_num.strip() ) - 1
+
+                for layer in data:
+                    if "LAYER:%i" % layer_num in layer:
```
For performance reasons it could be better to just not subtract the `- 1` like you do at line 69, but simply index the `data` array by the desired number. We guarantee that the layers match up, in order, with the layer numbers (1-indexed).

Then you do need to check for bounds though.

dmwatson replied:
I can rewrite it that way with some bounds checking. There's some things I was paranoid about in terms of cleaning the user input first and setting it automatically, but if it enhances performance I can rewrite it.

Ghostkeeper commented on scripts/PrusaColorChange.py:
```diff
@@ -0,0 +1,95 @@
+        # lines.insert( 2, "M600 ; Generated by PrusaColorChange plugin")
+        # Logger.log("d", "[PRUSACOLORCHANGE] Layer code: %s", lines[2] )
+        # final_line = "\n".join( lines )
+        # data[ layer_num ] = final_line
```
Please remove this debugging code.

dmwatson replied:
Not sure how that managed to stay in there past my debugging. I'll take that out.

Ghostkeeper commented:
It seems that M600 is supported by default in Marlin, so most printers should be able to make use of this.

dmwatson commented:
Based on what you mentioned about the M600 command being supported by Marlin firmware (and Smoothie and MK4duo apparently from my reading), would it make sense at this point to change the name of this plugin to something more generic like Color Change?

dmwatson committed:
Cleaned up PrusaColorChange issues

Cleaned up performance issue, default value for later retraction, and removed commented debugging code.

Ghostkeeper commented:
Yeah I think so.

dmwatson committed:
Fixed layer data adding back in.

Fixed adding layer data back into the correct indexed element in the data list.

jackha commented on scripts/PrusaColorChange.py:
```diff
@@ -0,0 +1,77 @@
+                    lines.insert(2, color_change )
+                    final_line = "\n".join( lines )
+                    data[ layer_num - 1 ] = final_line
+                    break
```
You may want to remove this "break" if we want more than 1 color change.

jackha commented:
I've added a comment. "break" should be removed or only a single color change is possible.
Also, a rename would be nice ^^

dmwatson committed:
Bugfix: removed extraneous break

Removed extraneous break and changed name from "Prusa Color Change" to "Color Change"

jackha commented:
Thanks for the fast reaction. However, can you rename the PrusaColorChange.py file to ColorChange.py as well? It doesn't work as is.

dmwatson committed:
File name change

Changed PrusaColorChange.py to ColorChange.py

dmwatson commented:
Sorry about that. Still getting used to GitHub and Git in general.

jackha commented:
Thanks, no problem :-)
